addons/l10n_id_efaktur/models/efaktur.py

```diff
@@ -2,57 +2,83 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from odoo import api, fields, models, _
-from odoo.exceptions import ValidationError
+from odoo.exceptions import UserError, ValidationError
 
 import re
 
 
 class Efaktur(models.Model):
     _name = "l10n_id_efaktur.efaktur.range"
     _description = "Available E-faktur range"
+    _rec_names_search = ["min", "max"]
 
     company_id = fields.Many2one('res.company', required=True, default=lambda self: self.env.company)
-    max = fields.Char(compute='_compute_default', store=True, readonly=False)
-    min = fields.Char(compute='_compute_default', store=True, readonly=False)
-    available = fields.Integer(compute='_compute_available', store=True)
+    max = fields.Char(required=True)
+    min = fields.Char(required=True)
+    available = fields.Integer(readonly=True)
+    next_num = fields.Integer(compute="_compute_next_num")
+
+    @api.depends('max', 'available')
+    def _compute_next_num(self):
+        for record in self:
+            record.next_num = int(record.max) - record.available + 1
+
+    def pop_number(self):
+        """ Consume the availability of a specific range to generate the eTax number for an invoice"""
+        self.ensure_one()
+
+        popped = self.next_num
+        self.available -= 1
 
-    @api.model
-    def pop_number(self, company_id):
-        range = self.search([('company_id', '=', company_id)], order="min ASC", limit=1)
-        if not range:
-            return None
-
-        popped = int(range.min)
-        if int(range.min) >= int(range.max):
-            range.unlink()
-        else:
-            range.min = '%013d' % (popped + 1)
         return popped
 
     @api.model
     def push_number(self, company_id, number):
-        return self.push_numbers(company_id, number, number)
-
-    @api.model
-    def push_numbers(self, company_id, min, max):
-        range_sup = self.search([('min', '=', '%013d' % (int(max) + 1))])
-        if range_sup:
-            range_sup.min = '%013d' % int(min)
-            max = range_sup.max
-
-        range_low = self.search([('max', '=', '%013d' % (int(max) - 1))])
-        if range_low:
-            range_sup.unlink()
-            range_low.max = '%013d' % int(max)
-
-        if not range_sup and not range_low:
+        """ Restoring the eTax number that got released after doing reset eFaktur on an invoice so
+        that it can be reused
+
+        :param company_id (int): company ID
+        :param number (str): number to be restored
+        """
+        number_int = int(number)
+        efaktur_range = self.search([('company_id', '=', company_id), ('min', '<=', number), ('max', '>=', number)], limit=1)
+
+        # if the released number is the last popped number from the range, we simply extend the availability
+        if efaktur_range.next_num == int(number) + 1:
```

Comment:
Hi @qdp-odoo , I tested this one and actually if we release the last number of a fully used range, it will add the `available` by 1, so I think we can proceed without changing this one!